### PR TITLE
async/await plus a variety of other changes

### PR DESCRIPTION
**Pull request by ggriffiniii**
I'm sending this at the risk of overwhelming as I recognize it's a pretty large pull request. This started with moving to std::future::Future as discussed in #111 then snowballed into a variety of changes (hopefully improvements) that could be made while we were already breaking semver compatibility.

Feel free to pick and choose which commits you like. The first commit is the only one necessary to get to std::future::Futures. As mentioned in the feature request this currently pins a few crates that are released with alpha labels. I would expect to wait for those dependencies to no longer be in alpha before releasing these changes.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -14,19 +14,18 @@ edition = "2018"
...
 rustls = "0.16"
 serde = "1.0"
 serde_json = "1.0"
 serde_derive = "1.0"
 url = "1"
-futures = "0.1"
-tokio-threadpool = "0.1"
-tokio = "0.1"
-tokio-timer = "0.2"
+futures-preview = "=0.3.0-alpha.19"
```
Comment (ghost) [inline Cargo.toml]: Hi @ggriffiniii atleast `futures-preview` are no longer alpha so I guess you could use `futures 0.3.1` instead.

Comment (ggriffiniii) [reply Cargo.toml]: We need to align the dependency versions with the versions hyper is using so that traits match. Once hyper releases a new version we can update accordingly.

**Comment by dermesser** [comment]
btw, I know that you are working on this (and am very happy about it!), I just currently lack the time to help more :)

File: Cargo.toml (hunk 1 above)
Comment (ctaggart) [reply Cargo.toml]: https://github.com/hyperium/hyper/pull/2007 is the PR for hyper

**Comment by ctaggart** [comment]
These changes look great @ggriffiniii. @dermesser, any plans to review & merge, or add @ggriffiniii as a collaborator? I upgraded a command line application of ours to rust async/.await and now I'm going to see if I can use this to get device auth working with Azure. It would be nice to see it land on `master`.

**Comment by ggriffiniii** [comment]
> These changes look great @ggriffiniii. @dermesser, any plans to review & merge, or add @ggriffiniii as a collaborator? I upgraded a command line application of ours to rust async/.await and now I'm going to see if I can use this to get device auth working with Azure. It would be nice to see it land on `master`.

Let me know if you run into any issues.

**Comment by ctaggart** [comment]
@ggriffiniii, the format of `tokenstorage.json` has changed. With or without the file, I get:
```
thread 'main' panicked at 'authenticator: Custom { kind: InvalidData, error: Error("invalid type: map, expected a sequence", line: 1, column: 0) }', src/libcore/result.rs:1187:5
```

I improved the error so that it says:
```
     Running `/home/cameron/github/yup-oauth2/target/debug/test-device`
thread 'main' panicked at 'authenticator: Custom { kind: InvalidData, error: "loading file tokenstorage.json: invalid type: map, expected a sequence at line 1 column 0" }', src/libcore/result.rs:1187:5
```

In storage.rs:
```
        let token_vec: Vec<JSONToken> = serde_json::from_slice(&contents)
            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, format!("loading file {}: {}", filename.display(), e)))?;
```

I'm working on a branch that merged your branch in. I'll create a new one and try again just to make sure.

**Comment by ggriffiniii** [comment]
Yes, my branch changes the on disk format of tokenstorage.json. I'm assuming you're seeing that error when attempting to read a file cretaed with an old version? Do you still get the error if you delete the old file and start fresh? I should have called that out more explicitly in the pull request summary.

**Comment by ctaggart** [comment]
On this branch with a clientsecret.json, but a tokenstorage.json removed, I get:
```
cameron@d5520:~/github/yup-oauth2/examples/test-device$ cargo run
   Compiling yup-oauth2 v3.1.1 (/home/cameron/github/yup-oauth2)
   Compiling test-device v0.1.0 (/home/cameron/github/yup-oauth2/examples/test-device)
    Finished dev [unoptimized + debuginfo] target(s) in 7.03s
     Running `/home/cameron/github/yup-oauth2/target/debug/test-device`
thread 'main' panicked at 'authenticator: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/libcore/result.rs:1187:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

**Comment by ggriffiniii** [comment]
> On this branch with a clientsecret.json, but a tokenstorage.json removed, I get:
> 
> ```
> cameron@d5520:~/github/yup-oauth2/examples/test-device$ cargo run
>    Compiling yup-oauth2 v3.1.1 (/home/cameron/github/yup-oauth2)
>    Compiling test-device v0.1.0 (/home/cameron/github/yup-oauth2/examples/test-device)
>     Finished dev [unoptimized + debuginfo] target(s) in 7.03s
>      Running `/home/cameron/github/yup-oauth2/target/debug/test-device`
> thread 'main' panicked at 'authenticator: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/libcore/result.rs:1187:5
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
> ```

Oh, that was a dumb mistake when refactoring. Pushed an update that should hopefully fix it.

**Comment by ctaggart** [comment]
Thank you! I confirmed it works with both Google and Azure.

**Comment by ghost** [comment]
Hi @ggriffiniii Thanks for your superb work :-),  I was using this pull request for Google STT it was working great till [d291ebb](https://github.com/dermesser/yup-oauth2/pull/112/commits/d291ebb02391ed5069ad789c4994be61e335a66c). After latest update, I am getting a minor error regarding `JSONToken` 

```rust
   async fn stream(&self, uri: Uri) {
    |  ______________________^
   | |         let secret = yup_oauth2::read_service_account_key(&self.credentials.service_account_file)
   | |             .await
   | |             .expect("Failed to read service account file");
...  |
   | |         }
   | |     }
   | |_____^ `std::rc::Rc<std::cell::RefCell<yup_oauth2::storage::JSONToken>>` cannot be sent between threads safely
    |
```
    = help: the trait `std::marker::Send` is not implemented for `std::rc::Rc<std::cell::RefCell<yup_oauth2::storage::JSONToken>>`

**Comment by ggriffiniii** [comment]
> Hi @ggriffiniii Thanks for your superb work :-), I was using this pull request for Google STT it was working great till [d291ebb](https://github.com/dermesser/yup-oauth2/pull/112/commits/d291ebb02391ed5069ad789c4994be61e335a66c). After latest update, I am getting a minor error regarding `JSONToken`
> 
> ```rust
>    async fn stream(&self, uri: Uri) {
>     |  ______________________^
>    | |         let secret = yup_oauth2::read_service_account_key(&self.credentials.service_account_file)
>    | |             .await
>    | |             .expect("Failed to read service account file");
> ...  |
>    | |         }
>    | |     }
>    | |_____^ `std::rc::Rc<std::cell::RefCell<yup_oauth2::storage::JSONToken>>` cannot be sent between threads safely
>     |
> ```
> 
> ```
> = help: the trait `std::marker::Send` is not implemented for `std::rc::Rc<std::cell::RefCell<yup_oauth2::storage::JSONToken>>`
> ```

Thanks for the report. Are you sure it was commit d291ebb? The reference to Rc<RefCell<JSONToken>> makes me suspect the most recent change 9dd49d4 would be the culprit. Do you have any more context around the code that's failing to compile. I'll take a look and see if I can figure out what's going on, but it's not clear from that snippet why the JSONToken needs to be send.

BTW. Thanks for testing these changes. I appreciate it.

**Comment by ggriffiniii** [comment]
Ok. I think b3b8b73 will fix your issue. I'll add some tests to ensure this doesn't happen again.

**Comment by ghost** [comment]
@ggriffiniii Thanks for looking into it and fixing the issue.

> Are you sure it was commit d291ebb? The reference to Rc<RefCell> makes me suspect the most recent change 9dd49d4 would be the culprit.

Actually, I meant to say that when I first pulled the source, latest revision was at [d291ebb](https://github.com/dermesser/yup-oauth2/pull/112/commits/d291ebb02391ed5069ad789c4994be61e335a66c), which was working fine. However, when I fetched again, the culprit code which you pointed out was somewhere in the last 10 commits which I hadn't looked into.

 I tested after your fix it is working now :-), Thanks.

**Comment by ghost** [comment]
One question, does it refresh service account token automatically once expired?

**Comment by ggriffiniii** [comment]
> One question, does it refresh service account token automatically once expired?
I believe it should.

**Comment by ghost** [comment]
@ggriffiniii I used a following method to retrieve access token:

```rust
    async fn get_access_token(&self) -> String {    
        let secret = yup_oauth2::read_service_account_key(&self.credentials.service_account_file)    
            .await    
            .expect("Failed to read service account file");    
    
        let auth = ServiceAccountAuthenticator::builder(secret)    
            .build()    
            .await    
            .expect("Invalid credentials");    
    
        let scopes = &[&self.credentials.scope];    
        let token = auth.token(scopes).await.expect("Couldn't get token");    
        token.access_token    
    } 
```
If I use this immediately it works fine. However, when I send the request to the server may be after 2 or 3 hours, I get an error from google for invalid token, I didn't close the application through out the session. So apparently it seems the token is not refreshing automatically. I may be missing something here. The token string should remain same if token is refreshed just before expiry, right? Also is refreshing of token behind any timer?

**Comment by ggriffiniii** [comment]
> I used a following method to retrieve access token:
> 
> ```rust
>     async fn get_access_token(&self) -> String {    
>         let secret = yup_oauth2::read_service_account_key(&self.credentials.service_account_file)    
>             .await    
>             .expect("Failed to read service account file");    
>     
>         let auth = ServiceAccountAuthenticator::builder(secret)    
>             .build()    
>             .await    
>             .expect("Invalid credentials");    
>     
>         let scopes = &[&self.credentials.scope];    
>         let token = auth.token(scopes).await.expect("Couldn't get token");    
>         token.access_token    
>     } 
> ```
> 
> If I use this immediately it works fine. However, when I send the request to the server may be after 2 or 3 hours, I get an error from google for invalid token, I didn't close the application through out the session. So apparently it seems the token is not refreshing automatically. I may be missing something here. The token string should remain same if token is refreshed just before expiry, right? Also is refreshing of token behind any timer?

@abdul-rehman0 Ok. I think there's a misunderstanding. Your code creates an authenticator, retrieves a token, and returns the access_token string (destroying the authenticator once the function returns). This is probably not what you what you want. The access token that is returned is typically only valid for a relatively short lifetime (the 2-3 hours that you experience). The authenticator is responsible for refreshing an access token but will only do so when the `token` method is invoked. When you invoke the `token` method it first checks to see if it has retrieved any tokens that are still valid for the scopes requested, if so it will return the cached access token. If the cached access token is expired (or about to expire) it will retrieve a new access token.

To summarize: you should create a single long-lived authenticator and whenever you need an access token you invoke the `token` method and use it. By doing this you'll ensure that the access token is valid.

Let me know if that makes sense and if that solves the expiration problem you're having.

**Comment by dermesser** [comment]
I am following along a bit here - it seems you are still productively working on it, but please let me know once you have it in a form where I can look to merge it! :)

**Comment by ggriffiniii** [comment]
> I am following along a bit here - it seems you are still productively working on it, but please let me know once you have it in a form where I can look to merge it! :)

Great to hear. I've been continually hacking away and the changes have definitely become more and more opinionated the more I've done, so don't hesitate to push back on things. In general my goal is to keep the tip of this branch as a ready-to-merge candidate, with the exception that the dependencies are still alpha releases. tokio recently released a real 0.2 release, so the only dependency we're still waiting for is hyper. Once that happens I'll ping this thread and say it's ready for a look.

**Comment by ghost** [comment]
> To summarize: you should create a single long-lived authenticator and whenever you need an access token you invoke the token method and use it. By doing this you'll ensure that the access token is valid.

> Let me know if that makes sense and if that solves the expiration problem you're having.

@ggriffiniii thanks for the clarification, this approach works for me I tested it :).

**Comment by jroesch** [comment]
I'm excited for this release as the version update here and Google APIs means we can drop the tokio 1.0 dependence and upgrade multiple services at work. It looks like hyper is releasing tonight: https://github.com/hyperium/hyper/pull/2061 (or real soon).

**Comment by ggriffiniii** [comment]
hyper 0.13 is released! Now the next hurdle is getting a hyper-rustls release that supports hyper 0.13 :(

There is a PR already in the works here: https://github.com/ctz/hyper-rustls/pull/96

**Comment by ctaggart** [comment]
`persist_tokens_to_disk` is saving the tokens to a file, but the file should not be readable to other users. Permissions are currently 644, but should be 600. Should I move this to a separate issue?

``` rust
    let authenticator = DeviceFlowAuthenticator::builder(app_secret)
        .persist_tokens_to_disk(token_path)
        .device_code_url(format!("https://login.microsoftonline.com/{}/oauth2/v2.0/devicecode", &aad_tenant_id))
        .grant_type("urn:ietf:params:oauth:grant-type:device_code")
        .build()
        .await?;
    let token = authenticator.token(scopes).await?;
```

**Comment by ggriffiniii** [comment]
I can fix that in this branch, but just to confirm this isn't a regression with my changes, correct? this exists in the current master as well?

Also do you know how to use https://doc.rust-lang.org/std/os/windows/fs/trait.OpenOptionsExt.html to set the permissions correctly on windows? I don't know windows well.

**Comment by ctaggart** [comment]
I don't believe this is a regression. I'm only interested in permissions on Linux & Mac.

**Comment by ggriffiniii** [comment]
The token file should now get created with more secure permissions on unix. If anybody wants to tell me how to do it on windows I will update it too, but I'm not particularly motivated to learn.

**Comment by dermesser** [comment]
just checking by to see how it is going! It sounds like this may take a bit while we're waiting for a new hyper-rustls? Let me know when you want any changes to be merged :)

**Comment by jroesch** [comment]
Does anyone know what hyper-rusttls is blocked on? it looks like they changes bumping the version to a point release got merged but there hasn't been a crates release yet.

**Comment by ggriffiniii** [comment]
> Does anyone know what hyper-rusttls is blocked on? it looks like they changes bumping the version to a point release got merged but there hasn't been a crates release yet.

I believe it was actually just released about 15 minutes ago. Now everything is just waiting on me to port the changes to hyper 0.13. I was surprised by the number of changes made between the final 0.13 and the prior alpha, but I think the additional changes in 0.13.1 can make it a pretty trivial transition now.

**Comment by ggriffiniii** [comment]
Okay, switched to the released version of hyper. I see I'm behind a couple commits on master. I'll rebase over those changes tomorrow, but otherwise I think this is ready for review.

**Comment by ggriffiniii** [comment]
I've rebased the PR on top of the latest master. It turns out the issue addressed by the latest commits was already handled as a byproduct of my refactor so there weren't any changes in the end result.

Just to restate what I said earlier. This is now a PR of 71 commits which I realize is rather huge. While it started as a mostly mechanical change, many of the later changes were much more opinionated and I would welcome some discussion on the pros and cons of each.

**Comment by dermesser** [comment]
I realize it is quite opinionated, but I see some other users here as
well. You all probably use this library more in real life than I do, so
I tend to trust your changes. For a more fundamental discussion I will
need to take a closer look first, which I hope is ok.
> [
> {
> "@context": "http://schema.org",
> "@type": "EmailMessage",
> "potentialAction": {
> "@type": "ViewAction",
> "target": "
> https://github.com/dermesser/yup-oauth2/pull/112?email_source=notifications\u0026email_token=AAH6PFZMVTJXJ34EP7RJVYDQZJL37A5CNFSM4JLA55HKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHG3L3A#issuecomment-567129580
> ",
> "url": "
> https://github.com/dermesser/yup-oauth2/pull/112?email_source=notifications\u0026email_token=AAH6PFZMVTJXJ34EP7RJVYDQZJL37A5CNFSM4JLA55HKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHG3L3A#issuecomment-567129580
> ",
> "name": "View Pull Request"
> },
> "description": "View this Pull Request on GitHub",
> "publisher": {
> "@type": "Organization",
> "name": "GitHub",
> "url": "https://github.com"
> }
> }
> ]

**Comment by ctaggart** [comment]
Happy new year! Any chance that we can get this merged to master in January?

**Comment by dermesser** [comment]
yes, we can! I will probably not have the time to review this very deeply, but it looks like you've done some great work. Do you think I should merge this as is?

**Comment by ctaggart** [comment]
@dermesser, I think you should merge it as is. @ggriffiniii  has done some great work. What do you think about adding him as a contributor and letting him merge it?

**Comment by dermesser** [comment]
> @dermesser, I think you should merge it as is. @ggriffiniii has done some great work. What do you think about adding him as a contributor and letting him merge it?

Good point! @ggriffiniii I added you as collaborator, meaning you should have push access now. This also means you can continue without being blocked on my response.

I will publish it to crates.io as version 4.0.0 once this is merged.

**Comment by obiesie** [comment]
@ggriffiniii great work here!! Any reason why you have opted to go with package private scope for some of the structs like Storage? It would make it more easier to extend if these were public

**Comment by ggriffiniii** [comment]
> @ggriffiniii great work here!! Any reason why you have opted to go with package private scope for some of the structs like Storage? It would make it more easier to extend if these were public

There were a few factors for why I opted for an enum rather than keeping the Storage trait.

* Traits that return futures require allocation and virtual dispatch. This seemed unfortunate for a caching layer that frequently will not require any I/O.
* I couldn't find anyway with the old API for users to actually provide their own Storage implementation. The builders didn't provide a way to provide your own storage it only gave the choice between in-memory or on-disk.

Given that Storage appeared to only be used for internal purposes an enum seemed like a better fit that also avoids requiring allocation for the future.

That said I would like to hear how making them public would make it easier to extend. Could you elaborate on your goals?

**Comment by obiesie** [comment]
I am looking into adding some code that can handle the google auth flow and from first looks it seems the privacy policy will make this a bit harder.

**Comment by ggriffiniii** [comment]
> I am looking into adding some code that can handle the google auth flow and from first looks it seems the privacy policy will make this a bit harder.

Sorry, I was asking for an elaboration to see how I might be able to change the code to address your needs. This description doesn't really explain what you are trying to do with enough detail to be helpful to the discussion I'm afraid. I'm open to adjusting things if it improves the user experience, but I would need to know how the current implementation falls short. It's not just that the types have private visibility. Making `Storage` publicly accessible would not allow you to extend it, nor would it allow you to hand a different implementation to the `Authenticator` without other changes so it's not very clear why you want it to be made visible.

I'm also not quite sure what "google auth flow" would mean in this context other than using the installed flow, device flow, or service account to get oauth2 tokens from google servers, which this crate should already support.

Again I'm open to making changes to address shortcomings, but I need you to elaborate on what you're trying to do so we can try and come up with a way to meet your goals.

**Comment by ernieturner** [comment]
FWIW, I just pointed my application to your fork @ggriffiniii and was able to get everything working after a little bit of rewriting. So I'd also vote to have this get merged as well. It seems like the Storage request should probably be done as a separate PR at this point. There's probably enough changes in this PR as is and I'm not sure adding to it is going to help.

Thanks so much for doing the work on this @ggriffiniii, this was one of the last libraries I've been waiting on to get moved over to async.

**Comment by ggriffiniii** [comment]
Ok. I pushed the changes into master.

@dermesser: I would suggest creating a 4.0.0-alpha.1 release before the real 4.0.0. Hopefully, that will get more eyes on the changes in case there's any changes that should happen before committing to the public api.

**Comment by ggriffiniii** [comment]
@dermesser: Any estimate when we may see a 4.0.0-alpha.1 released on crates.io?

**Comment by ggriffiniii** [comment]
> @dermesser: Any estimate when we may see a 4.0.0-alpha.1 released on crates.io?
Nevermind. I see it was released a week ago. Don't know how I missed that. Thanks!

**Comment by dermesser** [comment]
don't worry :) I also just noticed I had forgotten to push the version bump commit.

**Comment by dermesser** [comment]
> Ok. I pushed the changes into master.
> 
> @dermesser: I would suggest creating a 4.0.0-alpha.1 release before the real 4.0.0. Hopefully, that will get more eyes on the changes in case there's any changes that should happen before committing to the public api.

about this - I think I missed your message about this, but I am not worrying here. It's cheap to release this branch as 4.x and upgrade to 5.x later if it turns out there's something bad about it.
